Fix current values being overridden.

grokys opened the pull request:
## What does the pull request do?

At the beginning of a re-evaluation pass we mark all effective values as unset in order to signal that we've not yet found a value for them in the value frames. We do this in all cases except when they have a local value: in the case of a local value, the effective value _is_ the value and so value frames should only take effect when they're higher priority than the effective value. However when I added `SetCurrentValue` support, I neglected to change this logic to also take into account current values.

## Fixed issues

Fixes https://github.com/AvaloniaUI/Avalonia/issues/10655
Fixes https://github.com/AvaloniaUI/Avalonia/issues/10671

grokys committed:
Added failing test for #10655.

grokys committed:
Fix current values being overridden.

At the beginning of a re-evaluation pass we mark all effective values as unset in order to signal that we've not yet found a value for them in the value frames. We do this in all cases except when they have a local value: in the case of a local value, the effective value _is_ the value and so value frames should only take effect when they're higher priority than the effective value. However when I added `SetCurrentValue` support, I neglected to change this logic to also take into account current values.

Fixes #10655
Fixes #10671

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031983-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

maxkatz6 committed:
Merge branch 'master' into fixes/10655-setcurrentvalue-with-style-2

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031993-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

grokys commented:
@amwx got it, thanks. Seems my initial fix was a little naive, will add a fix for this to the PR.

grokys committed:
Add additional failing test for SetCurrentValue.

grokys committed:
Check for SetCurrentValue when reevaluating.

grokys committed:
Another failing test for SetCurrentValue.

grokys committed:
Ensure a changing value overrides current value.

When a binding produces a new value, ensure that it overwrites any value in place from `SetCurrentValue` if the priority is the same.

grokys committed:
Another failing test for SetCurrentValue.

grokys committed:
Move exit-early priority check.

Priority may not be cleared to `UnsetValue` in the presence of `SetCurrentValue` so do the check upfront and compare against the priority of the next frame instead of `Unset`.

grokys commented:
@amwx added a fix for that scenario too. Your testing is much appreciated, I will wait for the next round of problems you find ;)

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0032072-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

amwx commented:
Thanks @grokys. Issues fixed and no others popped up so at least to me this is good to go
